### ISSUE TITLE
How to change default limit=100 from sonarr/radarr requests ? ( for /api/all )

### ISSUE THREAD

**Issue by lrnv** [Jackett/Jackett#1661]
Hi !
I got the feeling that the "aggregate" function of jackett, wich allow to manage providers **only** in jackett and not N+1 times (if you got N clients of jackett - 3 times for me Sonarr+Radarr), doas give less results than individual searches. 

When performing a manual search in jackett, i got a lot of results than sonarr and radarr cant have. So my question is : Is there some limitating number of result per indexer than jackett can send throug torznab // that sonarr and radarr can receive ? That would explain why the aggregate search gave less results. But if i can tweak this value, that could be great too !! 

Eg, if i search "Greys anatomy S09" in jackett, i got more than 300 results, but in radarr only 100. Since i want this serie in French, and that the french trackers are last in the list, the 100 firsts results dont contain what sonarr is looking for and therefore it found nothing. 

You guys got i an idea of what i could do ? 

Jackett version:  latest pre-relaease
Mono version (if not using Windows): 5.0.1 

-------------------------------
EDIT : Ok it's sonarr faults, since the requests in sonarr's debug file looks like : 
`17-8-13 14:12:30.4|Debug|Torznab|Downloading Feed http://192.168.1.28:9117/torznab/all/api?t=tvsearch&cat=5030,5040&extended=1&apikey=(removed)&offset=0&limit=100`

And in radarr : 
`17-8-13 14:24:07.9|Debug|Torznab|Downloading Feed http://localhost:9117/torznab/all/api?t=movie&cat=2000,2010,2020,2030,2035,2040,2045,2050,2060&extended=1&apikey=(removed)&offset=0&limit=100&imdbid=0346491`

Do you know how i can fix this "limit=100" parametter from Sonarr/Radarr to make it infinity?

-------------------------------
EDIT 2 : 

Ok with hidden parameters in sonarr/radarr i've been able to setup a "&limit=5000" additional parameter to torznab request. 

but the "limit=100" is still passed, resulting in an error. Hw can i overpass this ? 

`17-8-13 14:50:02.2|Debug|Torznab|Downloading Feed http://192.168.1.28:9117/torznab/all/api?t=tvsearch&cat=5030,5040&extended=1&limit=5000&apikey=(removed)&offset=0&limit=100&q=Breaking%20Bad&season=2`

**Comment by hotio** [comment]
Sad to see that the Sonarr devs were so quick in shutting this more than valid point down. The fault lies sorta indeed with Sonarr/Radarr, although not easily fixed by a higher number, because this means they would also request 5000 entries when doing individual provider calls.

So the value would probably have to be user configurable...

As a workaround setting the RSS Sync to 10min could reduce the chance of missing something.

**Comment by lrnv** [comment]
No, setting the RSS sink to a smaller time value wont solve the issue because you'll get the same flow of answer (if you'r looking for something OLD, you can assume the flow will not change), and the limit point fixed in the jackett request will stay the same.

Do you think i should EDIT the first post to make it shorter / more accurate ? (first github post in jackett's github)

**Comment by hotio** [comment]
Did not say it would solve the problem, it could in some cases help prevent missing releases on rss sync, not for older manual searches

**Comment by Taloth** [comment]
@hotio I closed it because it's bad form to cross-post a question like this, because the Jackett team should be quite capable of explaining the issue, and posting it on the Sonarr github page an hour later because there wasn't get a response yet is just wrong and wastes our time.
His problem isn't limit, it's paging being tricky for aggregate feeds. limit simply is the size of a page, the offset param determines which page.
Further more: That 100 value is determined BY JACKETT ITSELF, hence it's most certainly not a Sonarr/Radarr issue.

**Comment by hotio** [comment]
Understood, thanks for the clarification.

**Comment by lrnv** [comment]
@Taloth i didnt know it was not cool to crosspost, in facts when i saw thar i **could** crosspost i did it because it had something to do with the 2 softs. Point. I didnt thaugth i'll put more workforce of volontary on m'y issue or what ever, and i'm not a terorist. My fault, i didnt know this github option was a no-no to use.

However, you'r saying it:s not this limit argument but a pagging issue with the aggregate feed ? I'll look into that

**Comment by chibidev** [comment]
@Taloth @lrnv it's not that simple. Sonarr/Radarr passes a limit=100 parameter in the torznab URI, which will always result in 100 matches or less. Jackett could ignore that parameter and provide additional results, however that would break the contract and a bad idea imho.
I don't know how result paging in Sonarr works, I've just never seen a call to my own server with offset other than 0. Perhaps there wasn't that many results, maybe @Taloth can elaborate (or eventually I'll look into it).
Back to the original question @lrnv posted: the aggregate feed is doing more work than just searching all indexers. It adds additional results if it thinks it's related to the query you posted, and it might filter out some if it thinks it's not what you're looking for. The manual search in Jackett works by simply asking each and every tracker you select, no filtering, so additional titles might be present.

Now in my head the issue is twofolds:
1. Pagination in the aggregate is tricky, because it needs to track the number each indexer provided, and since it's not guaranteed that at different times the same indexers will result, it's fairly hard to implement that. (Also, pagination in most of the indexers seem to be non-existent.)
2. When you configure multiple indexers in Sonarr, it will query each and every one of them passing a limit=100 parameter. When you have 5 configured indexers, this may result in 500 torrents. Yet when querying the aggregate, it is only searching one indexer (passing the limit parameter), hence it will result in 100 at max.

No. 1 is really hard to look into, but definitely worths some brainpower. However for a quicker solution we should focus more on no. 2.
As I've said before, no. 2 could be resolved in Jackett simply ignoring the limit parameter in the Aggregate (or pass it on to the members instead of limiting its own response), however as it breaks the contract of the API I wouldn't want to do that. However, as per Sonarr, the aggregate indexer shouldn't be any different than any other indexers, therefore treating it differently is a bad idea as well.
I think the easiest solution would be to have a field in Sonarr to manually edit the limit parameter (i.e. I can set it to 9999 if I want, or 0 if I don't want to pass a limit there), and Sonarr will take care of pagination (25/50/100 per page).
A probably more reasonable (however it requires more effort) solution would be to return the limits parameter during the capability discovery (i.e. returning something like <limits max="100" default="50"/> in the ?t=caps page) and Sonarr should make the query based on that. This way we can set the Aggregate to have a decent default value.
@Taloth what do you think?

**Comment by Taloth** [comment]
@chibidev Sonarr does the following:
- Use the 'default' specified in caps -> limits, as you already noted.
- Fetch a page using the specified limit.
- Optionally fetch additional pages if the last retrieved page has the full length as specified by the limit.
- It'll do so till:
  - when the last page wasn't the full length.
  - or a maximum of 30 pages (=30 api queries), which is just an arbitrary limit to prevent infinite api calls.
  - or when there are >= 1000 results (per indexer) already.

The 1000 is a limit I might be willing to increase, but I'll have to discuss it with markus since it could have a nasty performance impact. (Given that the feed is sorted by date, it would mean the most recent 1000 releases, which is normally enough)

If you wish to increase the limit per page then you should definitely adjust the caps result for the /api/all endpoint. You'll need to change that anyway if you want to advertise a different maximum page size.

Aggregates are definitely harder to properly implement coz you need to track the results as a 'session', keep it sorted _and_ fetch more upstream pages as needed. While returning sorted and limited pages back to Sonarr, but I believe it's doable.
(Personally I'm not a fan of aggregates coz it hides the 'which indexer' info, but that's not my problem.)

It's complexity depends on whether your upstream sources support paging. If they don't, you simply query them, and cache the fully sorted aggregate list for a couple of seconds so you can serve multiple offset=x queries. Rather easy, i'd say.
But if they do support paging then it gets complicated, you have to split it up in multiple stages and generate an aggregate incrementally. The process isn't very hard, but it does have an iterative loop to _accurately_ populate an next page while preserving remaining items and not resort to pre-fetching all the upstream pages, which is especially important for the rss feed.
If you're serious about having an aggregate in Jackett, then you should most definitely implement it properly. I can provide additional suggestions if you want.

**Comment by chibidev** [comment]
@Taloth thanks for the detailed answer, that helps a lot. If Sonarr indeed uses the limit specified in caps, then it's relatively easy to increase the number of results for the aggregate.
Back to the question of paging: I agree that normally it wouldn't be hard to do for an aggregate, however the problem is that paging is rarely supported by the indexers as of now. So first it needs to be somehow introduced to Jackett altogether before we can start doing something about the aggregate.
If you have ideas how to make Jackett better we're happy to hear. Just bear in mind: "properly" means different things to different people :)

**Comment by Taloth** [comment]
The logic for paging for the aggregate is the same as the paging for non-aggregates coz the page size could be different between the api and upstream source, and thus needs the same kind of 'caching'. Alternatively the proper page size needs to be advertised on the caps per indexer so they match the upstream source.

I was under the assumption that Jackett supported paging for most indexers since usually sites actually do have pages on their webui. But even if the upstream source does not support paging, you'll still need to do something for the aggregate if you want the aggregate behavior to be the same as putting all indexers in Sonarr separately. Increasing the limit does something to remedy that, of course.

But while Sonarr uses the caps for the default limit, other apps might not.
"properly", in this case, would mean in accordance to the newznab/torznab spec, which does allow arbitrary offsets and limits (limited by maximum alone) to be used by clients. This effectively means that Jackett should have a data accessor separately from the View (offset+limit).

Btw. now I think of it, iirc nzbHydra does similar things, might be worth looking at what they're doing.

**Comment by kaso17** [comment]
#975 and #250 are related issues.
For proper pagination support we'll definitely will need some kind of internal state storing/caching for many indexers.
Until someone implements this a simple workaround for the aggregate feed would be to make the currently hard coded limit of 100 configurable and set it to something high (>= 1000) for the aggregate feed.

**Comment by JigSawFr** [comment]
@kaso17 maybe we can use my caching mode on my indexers to implement something ?

**Comment by kaso17** [comment]
@JigSawFr Not sure which caching mode you're referring too? The one which is implemented by some indexers (indexer specific code)?

I see to general solutions:
Extend the torznab protocol to add a way of returning internal state information as part of the results and getting it back on follow up queries. That would allow jackett to stay stateless. But as the clients (sonarr, etc.) would have to support this I'm not a big fan of this solution.
The better solution (no torznab API changes required) is to introduce another abstraction layer. The torznab query would be sent to the abstraction layer, the abstraction layer then keeps calling the "getMoreResults(searchQuery, internalState)" method which would return another set (one page) of results from the tracker until the required number of results are reached or no more results are available.
The abstraction layer would then return the corresponding view (based on limit/offset) to the client and also cache the state (and results) for followup queries.

**Comment by lrnv** [comment]
It's amazing to see how fast you are taking care of the issue, thanks a lot guys !

If i understand clearly, wont the group PR that @chibidev work on solve the aggregate issue ? Thinking about juste a group consisting of all trackers. But maybe i didnt get the issue right ;)

**Comment by chibidev** [comment]
@lrnv not quite. At least not yet. I do plan on addressing this, maybe within that PR, however the grouping by itself will not resolve this.
(In essence the aggregate feed currently can be thought of as a non-configurable group that always hold all the configured trackers. The implementation behind them is basically the same.)

**Comment by chibidev** [comment]
@kaso17 we could use a cookie, a JWT token or the ASP.NET session ID to "identify" the request, but that might as well not be needed.
I was thinking that the "abstraction layer" can be an on-demand enumerator, which grabs one page and grabs the next one when the existing results are enumerated, and continues doing so until all results are enumerated. It would be fairly easy to store this enumerator since there's not much actual data to store so complexity is not an issue (theoretically, still need to measure of course). And since Sonarr/Radarr will post the same query but with different offset, we can pretty much use that to cache these enumerators (and unload them after they expire - e.g. for 2 minutes).
Once this is done for all the indexers it becomes fairly easy to implement it for the Aggregate as well, which will result in proper paging for all indexers. However it does take a fair bit of refactoring.

**Comment by kaso17** [comment]
aa157cfcb92f826c44ab472399cf4807b898714e adds support for the caps/limits element
69aa38c1ff217769a73afe816fafc072981e6beb increases the limit for the aggregate indexer to 1000 until proper pagination is supported.
@lrnv is 1000 a suitable limit for you or do you expect even more results?

**Comment by kaso17** [comment]
@chibidev the newznab/torznab document doesn't specify that the client have to support any kind of state keeping (cookies/session IDs) between requests. Expecting support for cookies seems to be a bad idea in my opinion. I think we'll have to cache the results based on the query parameters and maybe the remote IP.

**Comment by JigSawFr** [comment]
@kaso17 totale agree, in dev mode (only) for my French indexers, I support request caching based on search query. At the beginning it was designed to limit impact when developing definition, but we could create something based on it.

**Comment by lrnv** [comment]
@kazo17

a thousand is a lot, and more than enough. Sonarr grabed about 200 releases before i get to home the day my jackett downloaded your commit. Thanks a lot for that, it's a lot better and it'll do the job for the moment

**Comment by kaso17** [comment]
I'm closing this. 
Discussion regarding proper paging support should be continued in #975/#250.
